Add contouring

ianthomas23 opened the pull request:
This is a very early stage WIP to look at adding contouring to Bokeh. To try out this branch you will first need to `conda install -c conda-forge contourpy>=1` or `pip install contourpy>=1`.

There are two demos in the root directory, `contour_example.py` produces a static plot whereas `contour_example_serve.py` is a dynamic one that you run using `bokeh serve --show contour_example_serve.py`. It produces this:

https://user-images.githubusercontent.com/580326/155216360-b0a7c406-9e63-440e-8bb0-92a09b174745.mp4

Here there is a periodic callback that recalculates the `z` array that is used for contouring, passes that to `contourpy` to calculate the filled contours and contour lines that are returned as arrays of (x, y) points, and updates the `ColumnDataSource`s of the glyphs used to display the contours.

There is a new top-level function `figure.contour` that creates the glyphs. It creates filled contours and/or contour lines depending on what you ask for. Filled contours are created as a `MultiPolygons` object and lines as a `MultiLine` object. There are lower level functions:
1. `_filled_to_xs_and_ys` and `_lines_to_xs_and_ys`: convert the data returned by `contourpy` into a form suitable for Bokeh.
2. `contour_coords`: call `contourpy` to generate the contours and returns nested arrays of coordinates.
3. `contour_data_dicts`: wrap the above to create data dicts that can be used to update a `CDS`.
Everything is open to discussion, this is just a demo so that we have something concrete to talk about rather than a blank sheet of paper. I'll open a github discussion shortly...

(Edit: github discussion is https://github.com/bokeh/bokeh/discussions/12021).

bryevdv commented on contour_example_serve.py:
```diff
@@ -0,0 +1,29 @@
+
+def callback():
+    global timestep
+    timestep += 1
```
Just FYI see the relatively under-used functions in `bokeh.driving` to avoid explict global management like this, e.g. https://docs.bokeh.org/en/latest/docs/reference/driving.html#bokeh.driving.count

ianthomas23 replied:
Nice!

Also, I bet there are some really useful internal functions somewhere to help deal with scalar vs vector properties, I just haven't found them yet.

ianthomas23 commented on bokeh/plotting/contour.py:
```diff
@@ -0,0 +1,135 @@
+# General API
+#-----------------------------------------------------------------------------
+
+def contour_coords(x, y, z, levels, want_fill, want_line):
```
Sure.

bryevdv commented:
Reiterating the comment in the weekly meeting, I think a main decision is whether to do something along the lines of `GraphRenderer`, i.e. have a `ContourRenderer`.  I do think the new serialization work might afford some opportunities here, as well as learning some lessons from `GraphRenderer` to do things better. 

As far as rationale: the reason I think a `ContourRenderer` would be better for users is that the data for the contour lines and polys is not independent. We really should afford an API to allow updating all of this "atomically" in one go. If we just return two glyph renderers and call it done, then updating contours will always necessitate users updating two things, sequentially, and I just foresee issues and questions arising from that. I think it may also make render updates more prone to stuttering artifacts. 

So what might we do better with a `ContourRenderer`? 

* JS side manages internal (non-synced) `MultiPolygon` and `MultiLine` glyphs that it delegates to, when it is called to render. 
* Data for both contour lines and regions can be set in one go. e.g. we could have a struct type with two `ColumnData` sub-fields:
    ```python
    new_contour_data = ContourData(level_data=line_data_dict, region_data=poly_data_dict)

    contour_renderer.data = new_contour_data
    ```
    This would not *preclude* users from trying to update `contour_renderer.data.level_data` independently but we don't have to demonstrate that anywhere.

With something like this, the only things that are synced are standard and familial visual properties as well as the new contour data struct. 

Thoughts?

bryevdv commented:
@ianthomas23 FYI the issue you linked is: *[[BUG] export csv column order not retain](https://github.com/bokeh/bokeh/issues/12012#)*

ianthomas23 commented:
@bryevdv Sorry! That would have made the situation much more confusing.

mattpap commented on bokeh/models/renderers.py:
```diff
@@ -249,6 +252,25 @@ def add_decoration(self, marking: Marking, node: Literal["start", "middle", "end
         return decoration
 
+
+_DEFAULT_CONTOUR_LINE_RENDERER = lambda: GlyphRenderer(
```
This needs a rebase and use the new `InstanceDefault`.

mattpap commented on bokeh/plotting/contour.py:
```diff
@@ -17,23 +17,73 @@
 
-def contour_coords(x, y, z, levels, want_fill, want_line):
+def from_contour(
+    x: Optional[np.ndarray],
```
`np.ndarray` should be replaced with `npt.NDArray[float]`, as mypy doesn't like the former (`import numpy.typing as npt`). We prefer `x: T | None` instead of `x: Optional[T]`, because such arguments are rarely truly optional, especially if there's no default value provided.

ianthomas23 replied:
I've been using `Optional[T]` rather than `T | None` in other projects for backwards compatibility. It didn't even occur to me that the pipe would be OK here.

`x` and `y` can be truly optional eventually as they are optional in contourpy.

mattpap replied:
Pipe is OK because we use `from __future__ import annotations`, which makes annotations lazy, and thus we don't care whether Python supports `|` on types, only mypy. There is a caveat though. Type aliases have to use `Union[]`, as those are always evaluated eagerly (one of joys of static typing in Python).

ianthomas23 replied:
Thanks for the clarification.

mattpap commented on bokeh/plotting/contour.py:
```diff
@@ -17,23 +17,73 @@
+def from_contour(
+    x: Optional[np.ndarray],
+    y: Optional[np.ndarray],
+    z: Union[np.ndarray, np.ma.MaskedArray],
```
Perhaps `npt.ArrayLike` would be a better option (I'm not sure).

ianthomas23 replied:
`npt.ArrayLike` is what I would like to use, but it is only available from `numpy` 1.20 onwards.

mattpap replied:
We already use this across the codebase. Just put the import under `TYPE_CHECKING` (from `typing` module), and it won't affect the required run-time version of numpy (which currently is `>=1.11`).

mattpap commented on bokeh/plotting/contour.py:
```diff
@@ -17,23 +17,73 @@
+    y: Optional[np.ndarray],
+    z: Union[np.ndarray, np.ma.MaskedArray],
+    levels: Sequence[float],
+    fill_color=None,
```
Sadly we don't have a type alias for color.

ianthomas23 commented:
Trying out the `ContourRenderer.data` idea so that `contour_example_serve.py` now has
``` python
    new_contour_data = contour_data(x, y, z, levels, fill_color=fill_color, line_color=line_color)
    contour_renderer.data = new_contour_data
```
in the periodic callback, which is much neater than before.

Under the hood it is not elegant yet.  I'm just using a `dict` of `str` to `ColumnData` rather than a bespoke class for `ContourRenderer.data`, and a nasty interception in `ContourRenderer.__setattr__` to update the glyphs to use the new `ColumnData`. Really the `ColumnData` in `ContourRenderer.data` should be the same ones used by the glyphs, or cunning proxies, instead of this.

ianthomas23 commented:
Returning to continue with this. Now the user can specify the full set of fill and hatch visual properties to go to the filled contour (`MultiPolygons`) and line visual properties to go to the contour lines (`MultiLine`). There is a part-implemented `ContourColorBar` that shows all of the visual properties. This tries to make it as easy as possible for the user to get a colorbar, something like
``` python
contour_renderer = from_contour(x, y, z, levels, fill_color=... )
fig.renderers.append(contour_renderer)

colorbar = contour_renderer.color_bar()
fig.add_layout(colorbar, 'right')
```

Perhaps there could be a similar easy-to-use way to get this information into a legend instead. for those situations when a legend would be more appropriate.

Here is an example (from `contour_example.py`) showing fill, hatch and line properties in the contour plot and colorbar.

![contour_example](https://user-images.githubusercontent.com/580326/174491558-6a9ac5ce-e573-483d-a427-d8f645721d1d.png)

bryevdv commented:
@ianthomas23 This looks really fantastic. I think a legend option would be great, but I would also not hold up an initial merge just for that. What do you consider the minimal amount of work left until this could be merged?

My only comment is about:
```
contour_renderer = from_contour(x, y, z, levels, fill_color=... )
fig.renderers.append(contour_renderer)
``` 
We have never demonstrated manipulating `renderers` directly as a go-to technique. I think we should add a helper method to figure, along the lines of `hexbin`, etc. to keep consistent.

ianthomas23 commented:
Minimal amount of work until we could merge something:

* Refactor of `ColorBar` and `ContourColorBar` so that rather than the latter inheriting from the former, instead we have a common base class (`BaseColorBar`).
* Support calculating/rendering just lines or just filled contours, rather than both.
* API decisions, such as `figure.contour` etc.
* Add tests.

Docs with examples could be in a follow-on PR, relatively quickly.

This all *might* be possible for 3.0, depending on how far away that really is. :smile:

The big limitation at the moment is that you have to specify the contour levels and provide vector visual properties of the correct length. Eventually you'd want to ask for say approximately 10 contour levels and get sensible levels calculated for you. This needs the concept of do you want linear or log-spaced levels (or other?) and the new colorbar class needs to cope with this, and then we have the problem of how to specify vector visual properties of an initially unknown length. But I don't think any decisions we make now will make any of that unsolvable.

ianthomas23 commented:
Possible showcase example. I have managed to hold back on using dashed line and hatch patterns so far.

https://user-images.githubusercontent.com/580326/176267258-ff1fbe2f-bc68-4599-bfac-dc3a7f79f7e3.mp4

ianthomas23 commented:
This is ready for review now. It is functionally complete for simple use cases and the output looks good. I will add examples and docs in another PR soon, and start a github discussion containing a roadmap of what I plan to add in time.

Observations:
1. The two key functions are `figure.contour` (which calls `from_contour`) and `contour_data`. You always start with the former, the latter is for python-driven updates.
2. Args are `(x, y, z, levels, ...)`. Currently `z` and `levels` are compulsory (eventually it will just be `z`) and the others optional. But I didn't want to put `z` before the optional `x` and `y` as that feels wrong.
3. There is a refactor of the `ColorBar` code into `BaseColorBar`, `ColorBar` and `ContourColorBar`. All the layout and common drawing is in `BaseColorBar`, `ColorBar` works as before and takes a `ColorMapper`, `ContourColorBar` renders the fill and/or lines from a `ContourRenderer`. There were lots of ways of splitting up `ColorBar`, I've picked one which I think is low risk but has quite a few virtual functions.
4. There are lots of python tests but just one bokehjs visual test as this has to fake the contour calculation which occurs on the python side.
5. `ContourRenderer.__setattr__` is either cunning or hacky, depending on your point of view. :smile:

Here is example code to generate the animation above:
```python
from bokeh.driving import count
from bokeh.palettes import PiYG
from bokeh.plotting import figure, curdoc
from bokeh.plotting.contour import contour_data
import numpy as np

x, y = np.meshgrid(np.linspace(-1, 1, 41), np.linspace(-1, 1, 41))
nlevels = 11
levels = np.linspace(-1.0, 1.0, nlevels)

def get_z(timestep):
    z = np.zeros_like(x)

    f = 0.15
    rp = 0.4 + 0.08*np.cos(timestep*f)
    rm = 0.4 - 0.08*np.cos(timestep*f)
    m = 0.95
    amps  = [ m, m, -m, -m]
    xmids = [-0.4,  0.4, -0.4,  0.4]
    ymids = [-0.4,  0.4,  0.4, -0.4]
    rads  = [rp, rp, rm, rm]

    for amp, xmid, ymid, rad in zip(amps, xmids, ymids, rads):
        z += amp*np.exp( -( (x-xmid)**2 + (y-ymid)**2 )/rad**2 )

    return z

@count()
def callback(timestep):
    z = get_z(timestep)
    new_contour_data = contour_data(x, y, z, levels)
    contour_renderer.data = new_contour_data

fig = figure(width=600, height=450)
contour_renderer = fig.contour(x, y, get_z(0), levels, fill_color=PiYG, line_color="black", line_width=[1]*5 + [3] + [1]*5)
fig.renderers.append(contour_renderer)

colorbar = contour_renderer.color_bar()
fig.add_layout(colorbar, 'right')

doc = curdoc()
doc.add_periodic_callback(callback, 40)
doc.add_root(fig)
```

ianthomas23 commented on bokeh/models/contour_renderer.py:
```diff
@@ -0,0 +1,120 @@
+# Dev API
+#-----------------------------------------------------------------------------
+
+_DEFAULT_CONTOUR_LINE_RENDERER = lambda: GlyphRenderer(
```
I tried to use `InstanceDefault` here but in the end kept with this. This comment about `GraphRenderer` is relevant:
https://github.com/bokeh/bokeh/blob/2f617f55427e35bbfde531340b7dfadf2d3c3fd4/bokeh/models/renderers.py#L275-L280

ianthomas23 commented on bokeh/plotting/contour.py:
```diff
@@ -0,0 +1,278 @@
+# Dev API
+#-----------------------------------------------------------------------------
+
+def from_contour(
```
This is only ever called from `figure.contour`. It is public as far as this file is concerned, but should be considered private for bokeh as a whole. So maybe it should have a prefixed underscore?

ianthomas23 commented on bokehjs/src/lib/models/renderers/contour_renderer.ts:
```diff
@@ -0,0 +1,88 @@
+    this.prototype.default_view = ContourRendererView
+
+    this.define<ContourRenderer.Props>(({Array, Number, Ref}) => ({
+      fill_renderer: [ Ref(GlyphRenderer), () => new GlyphRenderer({glyph: new MultiPolygons(), data_source: new ColumnDataSource()}) ],
```
I needed to explicitly include the defaults here otherwise I get an error about different defaults in python and javascript. But similar code is not needed in `GraphRenderer` despite it being very similar?

bryevdv replied:
I'm not sure offhand @mattpap might know

mattpap replied:
See
https://github.com/bokeh/bokeh/blob/b85d0cd2a52d81d0a64404e0fcd6ff1446a236be/bokehjs/test/defaults/index.ts#L60-L62

But I would ask if you really need a default here, especially if there are higher level APIs that will be preferable to use by the user. If this is legitimate, then sure keep it, but don't force yourself to fill in defaults. Bokeh (3.0) will be perfectly happy with objects whose properties have unset values, until serialization time.

ianthomas23 replied:
Thanks for pointing that out. I'd prefer not to use a default so I will look into it further.

bryevdv commented on bokeh/plotting/contour.py:
```diff
@@ -0,0 +1,278 @@
+    from ..transform import ColorLike
+
+    ContourColor: Union[ColorLike, Sequence[ColorLike]]
+    ContourColorOrPalette: Union[ContourColor, Palette, PaletteCollection, ContourColor]
```
Generally put `TYPE_CHECKING` blocks  at the top, right under the imports. Type aliases can also now be typed:
```
ContourColor: TypeAlias = Union[ColorLike, Sequence[ColorLike]]
```
If these type aliases are ever going to be useful in any public APIs we should define them in a `.types` module or something that users can find f they need to integrate our code in their type checking.

ianthomas23 replied:
I've moved the `TYPE_CHECKING` block up towards the top.

Publicly exposing types for colors, etc, seems like a much larger issue, and I think this is one of the first attempts to type some of the visual properties. I haven't even typed contourpy yet, so I not sure I want to be at the leading edge of any of the type annotations work.

bryevdv commented on bokehjs/src/lib/models/annotations/color_bar.ts:
```diff
@@ -103,153 +94,68 @@ export class ColorBarView extends AnnotationView {
 
-    this._major_scale = (() => {
-      if (color_mapper instanceof LinearColorMapper)
+    switch (true) {
```
personally I find this construction obscure. Is there a tangible benefit over if/else cascade?

ianthomas23 replied:
No idea! This is code that was already in `color_bar.ts` before I refactored it into `base_color_bar.ts` and `color_bar.ts`. I don't like it either! I can change it whilst I am here if you'd like.

ianthomas23 replied:
I have changed all the `switch (true` JS blocks to `if`-`else` as they seem clearer to me as well.

mattpap replied:
In this case it's a matter of taste more than anything else. However, if you have no idea, then why change it? Especially that it's easy to miss this in the overall shuffling of code around The right thing to do, was to keep the original and start a discussion. Then we would be able to decide what's the best approach.

bryevdv replied:
@mattpap It's not merely style preference, it's about keeping the codebase accessible to many, and avoiding extremely obscure and unusual features unless there is a concrete justification to prefer them.

bryevdv commented on bokeh/models/contour_renderer.py:
```diff
@@ -0,0 +1,120 @@
+    Levels at which the contours are calculated.
+    """)
+
+    def __setattr__(self, name: str, value: Unknown) -> None:
```
Ah I was looking for an example and it was in the comments:
```
    new_contour_data = contour_data(x, y, z, levels)
    contour_renderer.data = new_contour_data
```
I think this dovetails with my suggestion to have `contour` data return a`dataclass`, then I think this could be accomplished with a property setter rather than a `__setattr__`?

ianthomas23 replied:
Done the `dataclass` and `set_data` approach.

bryevdv commented on bokeh/plotting/contour.py:
```diff
@@ -0,0 +1,278 @@
+    *,
+    want_fill: bool = True,
+    want_line: bool = True,
+) -> Tuple[Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
```
Same suggestion for a `dataclass` return instead of an ad-hoc tuple structure

ianthomas23 commented on .lgtm.yml:
```diff
@@ -12,6 +12,7 @@ extraction:
     python_setup:
       requirements_files: false
       requirements:
+        - contourpy
```
It is an important question. I suppose we'd have to consider the number of dependencies that are added and their potential to cause problems. Pure python dependencies are relatively safe as they are platform agnostic, those with compiled C extensions are more complicated (need platform-specific wheels and C runtime libraries) and those with compiled C++ extensions yet more complicated (need C++ runtime as well as C).

ContourPy only requires NumPy and a C++ runtime, as is available for all the platforms supported by NumPy.

I'll do an experiment to look at the dependencies:

1. Create an empty conda environment (Python 3.10 on Linux x86_64). This straightaway gives you C (`libgcc-ng` and `libgomp`) and C++ (`libstdcxx-ng`) runtimes, version 11.2.0. Names and versions of the runtimes will differ on different platforms.
2. `conda install -c conda-forge bokeh` adds 28 packages, 19 of which are compiled C (mostly `numpy` and image related) and none are compiled C++.
3. `conda install -c conda-forge contourpy` installs just one new package (`contourpy`) and updates the C and C++ runtimes to 12.1.0. This is the first compiled C++ library added. The runtime is only updated because it is a recent release, 3 weeks ago.
4. `conda install -c conda-forge pandas` installs 7 new packages, 2 of which include at least one compiled C++ library (`numexpr`, `pandas`) and 1 is compiled C (`bottleneck`).
5. Geopandas is split into `geopandas-base` and `geopandas` on conda but not PyPI (they recommend using conda because with pip it "can be a challenge to install"). `conda install -c conda-forge geopandas-base` installs 13 new packages, 8 are compiled C and 3 compiled C++ (`geos`, `proj`, `shapely`). This is where I start to get nervous.
6. `conda install -c conda-forge geopandas` (after `geopandas-base` installed) installs 102 packages. That is not a typo. I haven't counted the C/C++ libraries here, but some noteworthy packages are `fiona`, `gdal`, `matplotlib-base`, `scikit-learn`.

I would include `contourpy` as a compulsory dependency, and I would be happy to include `pandas` too. In the event of any problem with either I think/know we can guarantee a fast response. But I would defer to the opinions of others here as I have a vested interest. I would not support adding any form of `geopandas` as a compulsory dependency.

ianthomas23 replied:
FWIW I should also say that `contourpy` will be a compulsory dependency of `matplotlib` 3.6.0.

bryevdv replied:
> I would include contourpy as a compulsory dependency, and I would be happy to include pandas too. In the event of any problem with either I think/know we can guarantee a fast response. But I would defer to the opinions of others here as I have a vested interest. I would not support adding any form of geopandas as a compulsory dependency.

I concur with this.

ianthomas23 replied:
I will leave adding Pandas as a hard dependency for another PR.

ianthomas23 commented on bokeh/models/annotations/legends.py:
```diff
@@ -225,6 +212,52 @@ def __init__(self, *args, **kwargs) -> None:
+
+
+class ContourColorBar(BaseColorBar):
+    ''' Color bar used for contours.
```
Done. I've avoided mentioning iso-anything as I don't want to put anyone off if I can avoid it.

ianthomas23 commented on bokeh/models/contour_renderer.py:
```diff
@@ -0,0 +1,120 @@
+            line_renderer=self.line_renderer,
+            levels=self.levels,
+            ticker=FixedTicker(ticks=self.levels),
+        )
```
No, I can't image a scenario where it would be called more than once.

I wanted an easy way to create a `ContourColorBar` without the user having to worry about how it works (referencing the line and fill renderers). I thought a member function of `ContourRenderer` was a better approach to a standalone function, but I appreciate that this isn't typical in Bokeh so I am happy to do it some other way.

ianthomas23 commented on bokeh/plotting/_figure.py:
```diff
@@ -648,6 +649,55 @@ def graph(self, node_source, edge_source, layout_provider, **kwargs):
         self.renderers.append(graph_renderer)
         return graph_renderer
 
+    def contour(self, x=None, y=None, z=None, levels=[], **visuals):
```
I'll change the default value of `levels` to be `None`, as that it what it will be in the long run when you want the code to determine a sensible sequence of levels based on the data.

The types here now match those of `from_contour` as that is what is called to do all of the work.

ianthomas23 commented on bokeh/plotting/contour.py:
```diff
@@ -0,0 +1,278 @@
+        want_fill = False
+
+    if not want_fill and not want_line:
+        raise RuntimeError("Neither fill nor line requested in contour_data")
```
Yes, that would make it consistent with the other exceptions.

ianthomas23 commented:
I've addressed most but not all of the comments, hope to do more tomorrow.

ianthomas23 commented on bokeh/models/contour_renderer.py:
```diff
@@ -0,0 +1,120 @@
+            else:
+                self.line_renderer.data_source.data = dict(xs=[], ys=[], levels=[])
+        else:
+            super().__setattr__(name, value)
```
I have removed the `__setattr__` to use `set_data` instead.

ianthomas23 commented:
I have addressed all of the comments.

It is now using a full set of `dataclass`es between the various functions and in particular the top-level user-facing `dataclass` is `ContourData` that is returned from `contour_data()` and passed to `ContourRenderer.set_data()` rather than modifying `__setattr__`. The extra complexity here is that the component `FillData` and `LineData` dataclasses need to be converted to dicts so that extra fields (visual properties) can be added before they are put in `ColumnDataSource`s. Rather than using `dataclass.asdict`, which uses `deepcopy`, I have written a shallow copy version as otherwise we have extra unnecessary copies of potentially large numpy arrays. The shallow copy convert to dict code is as recommended in the docs https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict.

mattpap commented:
> Rather than using dataclass.asdict, which uses deepcopy, I have written a shallow copy version as otherwise we have extra unnecessary copies of potentially large numpy arrays.

You could use `dict(**bokeh.util.dataclass.entries(a_dataclass))`, which has the additional benefit of supporting actual optional values (not `None`) if needed.

ianthomas23 commented:
> You could use `dict(**bokeh.util.dataclass.entries(a_dataclass), **associated_visuals)`, which has the additional benefit of supporting actual optional values (not `None`) if needed.

Thanks, I will take a look.

ianthomas23 commented:
`dict(entries(data_object))` works for my simple case.

bryevdv commented:
LGTM modulo failing isort check.

bryevdv commented:
@mattpap do you need more time?

bryevdv commented:
@mattpap I will plan to merge this Monday morning (PST) unless there is a comment otherwise
